common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -1904,4 +1941,42 @@ long getPos() {
       return pos;
     }
   }
+
+  /**
+   * A class that wraps all the push-based shuffle service metrics.
+   */
+  static class PushMergeMetrics implements MetricSet {
+    // couldNotFindOpportunityResponses tracks how many times a shuffle block collided because
+    // of another block for the same reduce partition was being written
+    static final String NO_OPPORTUNITY_RESPONSES_METRIC = "couldNotFindOpportunityResponses";
+    // tooLateResponses tracks how many times a shuffle block push request is too late
+    static final String TOO_LATE_RESPONSES_METRIC = "tooLateResponses";
+    // pushedBytesWritten tracks the length of the pushed block data written to file in bytes
+    static final String PUSHED_BYTES_WRITTEN_METRIC = "pushedBytesWritten";
+    // cachedBlocksBytes tracks the size of the current deferred block parts buffered in memory.
+    static final String CACHED_BLOCKS_BYTES_METRIC = "cachedBlocksBytes";
```

Comment:
> Additionally, do we want to track the number of blocks which are deferred ?

Not sure if this would be a bit redundant, given we have tracked the deferred blocks in bytes, @otterc thoughts?

Reply:
There is a correlation between number of deferred blocks and amount of bytes deferred - but they are not necessarily the same, and also are influenced by the spark application configuration.

Reply:
The reason to track deferred bytes is because we want to know how much memory can be used up by these blocks. What insight can we derive from tracking the number of these blocks?

Reply:
As I mentioned above, they are related but not the same - particularly when applications do not have homogeneous configurations.

Reply:
Can we add `deferredBlocks` as well ? Which track the number of deferred blocks (in addition to the size you had added support for here) ?

Reply:
Thanks for the suggestion. Added `deferredBlocks`, PTAL.
WIP on adding UT.